File: leetcode-go/Q051-Q100/Q089_Gray_Code/Q089_Gray_Code.go
package Q089_Gray_Code

func grayCode(n int) []int {
	ans := make([]int, 1<<uint32(n))
	for i := 0; i < 1<<uint32(n); i++ {
		ans[i] = i ^ (i >> 1)
	}
	return ans
}

File: src/Y2018/backtracking/Q089_Gray_Code.java
package Y2018.backtracking;

import java.util.ArrayList;
import java.util.List;

public class Q089_Gray_Code {

    public List<Integer> grayCode(int n) {
        List<Integer> result = new ArrayList<>(1 << n);
        result.add(0);
        for (int i = 0; i < n; i++) {
            int size = result.size();
            for (int j = size - 1; j >= 0; j--) {
                result.add(result.get(j) | 1 << i);
            }
        }
        return result;
    }

}
